### ISSUE TITLE
Enemies don't attack

### ISSUE THREAD

**Issue by Chance4us** [diasurgical/devilutionX#2232]
For some unknown reason, the enemies don't attack and just stand around.
It looks strange, although the attached savegame also behaves in the same kind in vanilla. 

![screen01](https://user-images.githubusercontent.com/45902449/123512161-3ea69b80-d686-11eb-95a7-fb7bd2f6746e.png)
[single_0.hsv.zip](https://github.com/diasurgical/devilutionX/files/6720047/single_0.hsv.zip)

**Comment by obligaron** [comment]
I did a first look at the code:

- the pack Monsters are only allowed to be [max 4 tiles away from leader future position](https://github.com/diasurgical/devilutionX/blob/fb235212882b3a2ea506d74394a01fc7dd181a98/Source/monster.cpp#L4671)
- It seems that the leader of the pack doesn't target you, cause not all monsters of his pack are near him and he [tries to stick with them](https://github.com/diasurgical/devilutionX/blob/fb235212882b3a2ea506d74394a01fc7dd181a98/Source/monster.cpp#L4695)

But this was only a first look and someone with more knowledge of the AI should take a look. 🙂

**Comment by AJenbo** [comment]
Yeah I knew already from just seeing the screen shot that it's a pack tied to a boss and the group is having trouble moving as a whole, probably some are stuck around the corner. It's a fairly common thing. I'm a bit hesitant to label this as a bug since changing it could be pretty drastic to the way AIs are allowed to move.

**Comment by FitzRoyX** [comment]
> * the pack Monsters are only allowed to be [max 4 tiles away from leader future position]

That probably wasn't the best idea in a game full of tight spaces. You definitely want enemies to aggro and pursue, not be fish in a barrel.

**Comment by StephenCWills** [comment]
> That probably wasn't the best idea in a game full of tight spaces. You definitely want enemies to aggro and pursue, not be fish in a barrel.

I guess part of the problem is you want these packs to be strong and scary, barreling down on the hero in a large group rather than allowing them to be separated and picked off one-by-one.

**Comment by Chance4us** [comment]
In any case, they should also be aggressive as a pack.

**Comment by FitzRoyX** [comment]
> > That probably wasn't the best idea in a game full of tight spaces. You definitely want enemies to aggro and pursue, not be fish in a barrel.
> 
> I guess part of the problem is you want these packs to be strong and scary, barreling down on the hero in a large group rather than allowing them to be separated and picked off one-by-one.

Does bumping the limit from 4 to 10 decongest the group and get it moving? I don't know how else you stop this without massively raising or removing the limit. Sometimes the pack leader is going to be behind his units on approach, as he is here.

**Comment by AJenbo** [comment]
I don't think that would solve it but just move the issue to a different situation, they would probably have to be more intelligent about what they are doing as a group and if they are indeed in an effective group formation more then just a distance check.

**Comment by julealgon** [comment]
@nitekat wanna share your insights on this one? It seems to be about that "boss leash" issue that happens quite often with balrogs.

**Comment by NiteKat** [comment]
This issue is also common for boss packs that can charge. There's a goatman group with charge, a bat group with charge, and the Horned Demon type packs. If a single one of the group charges, they are immediately out of the leash range in most cases, and so the pack just freezes up, only occasionally charging if the conditions are right.

It's a tough one to say how it should be fixed... Maybe the most vanilla option is to put some logic for them to do when they are outside of the leash range to try to get the group moving again instead of just getting frozen? Like either find a median point they can all try to move to in order to regroup, or the monsters that are outside the leash range return to the group to be in leash range again?

Even those solutions could be a problem though in a tight space where the unique monster ended up in the back.

Maybe another possibility is to have the leashed monsters, instead of requiring being within certain distance, they share squelch and player position information? So that way if you wake one up, they all wake up, and if you somehow separate one from the pack, the rest still know where the player is and won't calm down and will come for the player?

**Comment by AJenbo** [comment]
From looking at `GroupUnity()` it appears that when groups get split up they are suppose to designate a new group leader for the run off, but the code related to it appears very messy.

**Comment by AJenbo** [comment]
@Chance4us I'm not expecting any miracles here, but could you check if there has been any change to this in the latest test builds?

**Comment by NiteKat** [comment]
So I gathered a couple of saves where, according to @AJenbo , the monsters are still stuck even with https://github.com/diasurgical/devilutionX/pull/3118 applied (I didn't have the time/setup to test with this PR myself):
[single_1.zip](https://github.com/diasurgical/devilutionX/files/7349929/single_1.zip) <-- Group of big guys that are stuck in a catacombs doorway

[single_1.zip](https://github.com/diasurgical/devilutionX/files/7349931/single_1.zip) <-- Solo scavenger unique that is stuck.

**Comment by Chance4us** [comment]
Tested both attache savegame files too.
Conclusion "Solo scavenger unique that is stuck.":
I suspect it's a different issue because all of the group members are dead and only the leader is stuck and can not move.
The leader stands very close to the wall and there is at least the possibility that he was placed in a bad position so that he is a little way into the wall. There is an impression that he is simply unable to move from his position, even if he attacks.

Conclusion "Group of big guys that are stuck in a catacombs doorway.":
Here it is actually the case that as soon as the game is loaded, the whole group is still stuck.
But as soon as the hero moves a little in their direction and an enemy from the group starts to attack, the group begins to unstuck.
This is a positive improvement because this does not happen before the pull request.

All in all, the pull request improves the behavior of the game.

**Comment by Chance4us** [comment]
> > Likely the scavenger has ended up with a none zero packsize despite all members of the pack being killed. I'll try and see if I can fix that.
> 
> This may be a different issue altogether than, that is unique to the uniques in the scavenger family, as they get stuck like this quite often.

That's exactly what I think.

**Comment by Chance4us** [comment]
> If we change it so that the monster can make the evaluation when idling, this lets the AI continue instead of falling in to this state:

It would make sense because the group would behave as the players would expect to be attacked.

**Comment by Chance4us** [comment]
@AJenbo I missed your invitation to support. If you need further special tests or bisecting, you can leave me a message at my email address for Github activities. This is Chance4us@yandex.com
Please note that I have to work a lot and also spend a lot of time doing sports. But if there is some meticulous work to be done, I can support you with my remaining free time.

**Comment by Chance4us** [comment]
@AJenbo I found next another frozen specialist.
![screen00](https://user-images.githubusercontent.com/45902449/137814455-1648872f-9bf3-4ff6-ac29-d349eb60566a.png)
[spineeater.zip](https://github.com/diasurgical/devilutionX/files/7368826/spineeater.zip)
